src/Parser.js

```diff
@@ -127,11 +114,16 @@ Parser.prototype.parseInput = function(pos, mode) {
  */
 Parser.prototype.parseExpression = function(pos, mode, breakOnInfix, breakOnToken) {
     var body = [];
+    var lex = null;
     // Keep adding atoms to the body until we can't parse any more atoms (either
     // we reached the end, a }, or a \right)
     while (true) {
-        var lex = this.lexer.lex(pos, mode);
-        if (breakOnToken != null && lex.text === breakOnToken) {
+        lex = this.lexer.lex(pos, mode);
+        if (["}", "\\end", "\\right", "&", "\\\\", "\\cr"]
+            .indexOf(lex.text) !== -1) {
+            break;
+        }
+        if (breakOnToken !== null && lex.text === breakOnToken) {
```

Comment:
If you make this `breakOnToken != null`, then you can just make the last argument optional instead of having to pass `null` in everywhere.